### DISCUSSION TITLE
Retrieve fixture directly

### DISCUSSION THREAD

**Discussion by effigies** [pytest-dev/pytest#10223] [Q&A] [answered]
I have a somewhat odd use case in that I'm working in a test suite that already has its own parameterization (it was written back in the nose days). Basically you write code like:

```Python
class MySuite(APITests):
    def validate_X(self, objmaker, params):
        myobj = objmaker()
        # write tests based on params
```

The `APITests` superclass will then turn `validate_X` into a `test_X` function that loops over `(objmaker, param)` tuples, which the runner like `nose` or `pytest` will happily pick up.

Annoyingly, this makes it impossible to use pytest fixtures in this way:

```Python
class MySuite(APITests):
    def validate_X(self, objmaker, params, tmp_path):
        myobj = objmaker()
        # write tests based on params
```

Is it possible to retrieve an active fixture by name, e.g.:

```Python
@pytest.mark.usefixtures('tmp_path')
class MySuite(APITests):
    def validate_X(self, objmaker, params):
        tmp_path = pytest.fetch_fixture('tmp_path')
        myobj = objmaker()
        # write tests based on params
```

An even better approach, IMO, would be to use it as a context manager that does setup and teardown internal to the function:

```Python
class MySuite(APITests):
    def validate_X(self, objmaker, params):
        with pytest.fetch_fixture('tmp_path') as tmp_path:
            myobj = objmaker()
            # write tests based on params
```

(I'm using `tmp_path` as an example. That one is quite easy to make my own context manager for, but there are more complicated fixtures that it would be nice to take advantage of in this context.)

Apologies if this is a duplicate. The closest I could find was https://github.com/pytest-dev/pytest/issues/3834. Having something like the latter would probably resolve https://github.com/pytest-dev/pytest/issues/2703 as well.

**Comment by RonnyPfannschmidt** [comment]
What you suggest is intentionally unsupported

The only way to add it would be  global state, which is a unacceptable footgun

**Comment by effigies** [comment]
Well, I'd hoped there would be some way, but thanks for the prompt reply.

**Comment by The-Compiler** [comment, marked as answer]
Couldn't you just have an autoused fixture which stores the fixture objects in the instance of the test object? Something like:

```python
class MySuite(APITests):
    @pytest.fixture(autouse=True)
    def setup(self, tmp_path):
        self.tmp_path = tmp_path

    def validate_X(self, objmaker, params):
        tmp_path = self.tmp_path
        myobj = objmaker()
        # write tests based on params
```

**Reply by effigies** [reply]
Nice! I'd tried something similar but hadn't quite landed here. Thanks!
